File: c++/array/q152.cpp
int maxProduct(vector<int>& nums) {
    if(nums.size()==0)
        return 0;
    int r = nums[0];
    int imax = r;
    int imin = r;
    for(int i = 1; i < nums.size(); i++){
        if(nums[i] < 0)
            swap(imax,imin);

        imax = max(nums[i],imax*nums[i]);
        imin = min(nums[i],imin*nums[i]);
        r = max(imax,r);
    }

    return r;
}

File: python/array/q152.py
"""
Find the contiguous subarray within an array (containing at least one number) which has the largest product.

For example, given the array [2,3,-2,4],
the contiguous subarray [2,3] has the largest product = 6.
"""
def maxProduct(self, nums):
    """
    :type nums: List[int]
    :rtype: int
    """
    if len(nums)==0:
        return 0;
    r = imax = imin = nums[0]
    for val in nums[1:]:
        if val < 0:
            imax,imin = imin,imax
        imax = max(val,val*imax)
        imin = min(val,val*imin)
        r = max(r,imax)
    return r
